src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -535,26 +616,39 @@ private void InheritNullableStateOfFieldOrProperty(int targetContainerSlot, int
             }
         }
 
-        private void InheritNullableStateOfAnonymousTypeInstance(TypeSymbol targetType, int targetSlot, int valueSlot, bool isByRefTarget)
+        private void InheritDefaultState(int targetSlot)
         {
             Debug.Assert(targetSlot > 0);
-            Debug.Assert(targetType.IsAnonymousType && targetType.IsClassType());
 
-            foreach (var member in targetType.GetMembersUnordered())
+            // Reset the state of any members of the target.
+            for (int slot = targetSlot + 1; slot < nextVariableSlot; slot++)
             {
-                if (member.Kind != SymbolKind.Property)
+                var variable = variableBySlot[slot];
+                if (variable.ContainingSlot != targetSlot)
                 {
                     continue;
                 }
+                this.State[slot] = !variable.Symbol.GetTypeOrReturnType().IsNullable;
+                InheritDefaultState(slot);
+            }
+        }
 
-                var propertySymbol = (PropertySymbol)member;
+        private void InheritNullableStateOfTrackableType(int targetSlot, int valueSlot, bool isByRefTarget)
+        {
```

Comment:
It feels like this copying of state could result in an infinite loop in trying to converge on a final state, for example if the program contains a loop, each pass of analysis would track longer and longer chains of dotted names.

Reply:
We're only copying nullable state for members of reference types when assigning object creation expressions or anonymous type instances currently. We'll need to decide what other cases should be handled here.